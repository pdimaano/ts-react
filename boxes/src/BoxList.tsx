import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

/** Manage list of boxes
 *
 * State:
 * - boxes: [ { id, width, height, backgroundColor }, ... ]
 *
 *  App -> BoxList -> NewBoxForm and Box
 */

interface BoxInterface {
  id: number;
  width: number;
  height: number;
  backgroundColor: string;
}

function BoxList() {
  const [boxes, setBoxes] = useState<BoxInterface[]>([])

  /** add box with given { id, width, height, backgroundColor } */
  function add(newBox: BoxInterface) {
    setBoxes(boxes => [...boxes, newBox]);
  }

  /** remove box matching that id. */
  function remove(id: number) {
    setBoxes(boxes => boxes.filter(box => box.id !== id));
  }

  return (
    <div>
      <NewBoxForm createBox={add} />
      {boxes.map(({ id, width, height, backgroundColor }) => (
        <Box
          key={id}
          id={id}
          width={width}
          height={height}
          remove={remove}
          backgroundColor={backgroundColor}
        />
      ))}
    </div>
  );
}

export default BoxList;
