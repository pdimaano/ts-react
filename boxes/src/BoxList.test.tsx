import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";
import Box from "./Box";

/** Adding a box for testing */
function addBox(container: HTMLElement, height = 2, width = 3, color = "peachpuff") {
    const heightInput = container.querySelector("[name=height]");
    const widthInput = container.querySelector("[name=width]");
    const backgroundInput = container.querySelector("[name=backgroundColor]");

    fireEvent.change(backgroundInput!, { target: { value: color } });
    fireEvent.change(widthInput!, { target: { value: width } });
    fireEvent.change(heightInput!, { target: { value: height } });

    // test button to make sure it works
    const button = container.querySelector(".NewBoxForm-addBtn");
    fireEvent.click(button!);
}

it("matches snapshot when no boxes", function () {
    const { container } = render(<BoxList />);
    expect(container).toMatchSnapshot();
});


describe("adding boxes", function () {
    it("can add a new box", function () {
        const { container } = render(<BoxList />);

        // no boxes yet
        expect(container.querySelector(".Box")).not.toBeInTheDocument();
        addBox(container);

        // expect to see a box
        const box = container.querySelector(".Box");
        expect(box).toBeInTheDocument();
        expect(box!.querySelector(".Box-box")).toHaveStyle(`
        width: 3em;
        height: 2em;
        background-color: peachpuff;
        `);
    });
})