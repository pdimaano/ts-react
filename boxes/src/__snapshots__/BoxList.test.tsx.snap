// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding boxes matches snapshot after adding box 1`] = `
<div>
  <div>
    <div>
      <form>
        <div>
          <label
            for="newBox-height"
          >
            Height
          </label>
          <input
            id="newBox-height"
            name="height"
            value=""
          />
        </div>
        <div>
          <label
            for="newBox-width"
          >
            Width
          </label>
          <input
            id="newBox-width"
            name="width"
            value=""
          />
        </div>
        <div>
          <label
            for="newBox-backgroundColor"
          >
            Background Color
          </label>
          <input
            id="newBox-backgroundColor"
            name="backgroundColor"
            value=""
          />
        </div>
        <button
          class="NewBoxForm-addBtn"
        >
          Add a new box!
        </button>
      </form>
    </div>
    <div
      class="Box"
    >
      <div
        class="Box-box"
        style="height: 2em; width: 3em; background-color: peachpuff;"
      />
      <button
        class="Box-removeBtn"
      >
        Remove The Box!
      </button>
    </div>
  </div>
</div>
`;

exports[`matches snapshot when no boxes 1`] = `
<div>
  <div>
    <div>
      <form>
        <div>
          <label
            for="newBox-height"
          >
            Height
          </label>
          <input
            id="newBox-height"
            name="height"
            value=""
          />
        </div>
        <div>
          <label
            for="newBox-width"
          >
            Width
          </label>
          <input
            id="newBox-width"
            name="width"
            value=""
          />
        </div>
        <div>
          <label
            for="newBox-backgroundColor"
          >
            Background Color
          </label>
          <input
            id="newBox-backgroundColor"
            name="backgroundColor"
            value=""
          />
        </div>
        <button
          class="NewBoxForm-addBtn"
        >
          Add a new box!
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`removing boxes matches snapshot after removing box 1`] = `
<div>
  <div>
    <div>
      <form>
        <div>
          <label
            for="newBox-height"
          >
            Height
          </label>
          <input
            id="newBox-height"
            name="height"
            value=""
          />
        </div>
        <div>
          <label
            for="newBox-width"
          >
            Width
          </label>
          <input
            id="newBox-width"
            name="width"
            value=""
          />
        </div>
        <div>
          <label
            for="newBox-backgroundColor"
          >
            Background Color
          </label>
          <input
            id="newBox-backgroundColor"
            name="backgroundColor"
            value=""
          />
        </div>
        <button
          class="NewBoxForm-addBtn"
        >
          Add a new box!
        </button>
      </form>
    </div>
  </div>
</div>
`;
